shared/js/background/events.es6.js

```diff
@@ -77,54 +77,61 @@ browser.runtime.onInstalled.addListener(onInstalled)
  * ONBOARDING
  * Logic to allow the SERP to display onboarding UI
  */
-
-let onBeforeNavigateTimeStamp = null
-browser.webNavigation.onBeforeNavigate.addListener(details => {
-    onBeforeNavigateTimeStamp = details.timeStamp
-})
-
 browser.webNavigation.onCommitted.addListener(async details => {
     await settings.ready()
     const showWelcomeBanner = settings.getSetting('showWelcomeBanner')
     const showCounterMessaging = settings.getSetting('showCounterMessaging')
 
-    // We show the welcome banner and counter messaging only once
+    // We show the welcome banner and counter messaging only once, and we only
+    // show the counter messaging the first time a search navigation is seen.
     if (showWelcomeBanner || showCounterMessaging) {
         const isAddressBarQuery = details.transitionQualifiers.includes('from_address_bar')
 
-        if (showWelcomeBanner) {
-            settings.removeSetting('showWelcomeBanner')
-        }
         if (isAddressBarQuery && showCounterMessaging) {
             settings.removeSetting('showCounterMessaging')
         }
 
-        if (onBeforeNavigateTimeStamp < details.timeStamp) {
-            if (browserName === 'chrome') {
-                browserWrapper.executeScript({
-                    target: { tabId: details.tabId },
-                    func: onboarding.onDocumentStart,
-                    args: [{
-                        duckDuckGoSerpHostname: constants.duckDuckGoSerpHostname
-                    }],
-                    injectImmediately: true
-                })
-            }
+        // Avoid displaying the welcome banner if the tab isn't active, or
+        // another navigation has started.
+        let {
+            url: tabUrl, active: tabActive
+        } = await browser.tabs.get(details.tabId)
+        tabUrl = new URL(tabUrl)
+        if (!tabActive ||
+            tabUrl.hostname !== constants.duckDuckGoSerpHostname ||
+            tabUrl.pathname !== '/' ||
+            !tabUrl.searchParams.get('q')) {
+            return
+        }
 
+        if (showWelcomeBanner) {
+            settings.removeSetting('showWelcomeBanner')
+        }
```

Comment:
For some reasons, it seems that the `onCommitted` listener can be called twice.
As a result, in case of the counter-messaging, not removing the setting before the early return can cause the banner to be displayed twice.
Testing the counter-messaging is a bit annoying but can be done with the following steps:

- On the SERP (in `src/js/pages.SERP.js`), modify  `window.onFirstSearchPostExtensionInstall` so that the counter-messaging shows on your testing device. For instance, if you are on a Mac you can do:
    ```diff
       (
    -     (env.device.isOSX && env.device.chromeVersion >= 87 && env.device.chromeVersion <= 89) ||
    +     (env.device.isOSX && env.device.chromeVersion >= 87 /*&& env.device.chromeVersion <= 89*/) ||
          (env.device.isWindows && env.device.chromeVersion >= 87)
        )
    ```
- Rebuild the SERP (`npm i` `grunt` `r index`)
- Edit the extension code so that a search from the address point will hit your SERP dev instance e.g.
    - In constants.js set `duckDuckGoSerpHostname` to your dev instance
    - In manifest.json set `search_url` and `suggest_url` to your dev instance
    - Rebuild the extension
- Open Chrome
- Visit `https://<you>.duckduckgo.com`
- Clear the data (cache, cookies, local storage etc.)
- Uninstall the extension
- Quit Chrome (this is required so that the chrome message shows)
- Open Chrome
- Visit `https://<you>.duckduckgo.com` (this is required to handle auth)
- In another tab, visit [chrome://extensions](chrome://extensions)
- Install the patched version of the extension
- Due a search from the address bar (be sure that it's a search and not access to your history)
- [ ] Check that you see chrome message
- [ ] Check that you see the counter-messaging
- Click on "Keep it"
- [ ] 🐛  Check that you see 2 welcome messages displayed (this can be fixed by removing the `showWelcomeBanner` setting to before the early return, or, understanding why the callback is called twice).

Reply:
Looking more, it seems that the callback is called twice because the URL filter contains 2 entries (so it seems that Chrome register as many event listeners as there are entries in the URL filter option). I didn't investigate in details but this [old comment on a bug report](https://bugs.chromium.org/p/chromium/issues/detail?id=450068#c6) seems to point at smtg similar.

Reply:
> in case of the counter-messaging, not removing the setting before the early return can cause the banner to be displayed twice.

With this change we're removing the welcome banner setting after the early return, but still removing the counter-messaging setting every time. I don't think this change should affect the counter-messaging banner.

Reply:
@kzar apologies if i wasn't clear, If you follow the testing instructions above you will see 2 welcome banners rendered _on the same page_  after clicking on "keep it".
This is a regression introduced by this change.

Reply:
@kzar here is a screenshot of the issue (notice the 2 welcome banners)

<img width="1449" alt="Screen Shot 2022-07-06 at 9 27 30 AM" src="https://user-images.githubusercontent.com/1141327/177600137-3de063f2-1a7e-436d-b389-10a2ae8b0177.png">

I can confirm that the fact that the callback is called twice is the root cause (and related to the URL filter).
I think that the issue is not the early return per se but the fact that you have an async operation before (`await browser.tabs.get(details.tabId)`).
The way I understand it, both callbacks are called on the same tick, so when they are called they both see the value for `showWelcomeBanner` to be `true` hence the bug.

Reply:
Right OK, got it... I think!

I also can see our `onComitted` event listener firing twice like you said. I don't think it's due to the extra URL filter however, since I did a test with only one filter and it still fired twice. I suspect it's due to the request being redirected (I see the `TransitionQualifier` contains "server_redirect"), but that's just a guess.

The way I figure it, this `onBeforeNavigateTimeStamp` workaround is therefore serving to avoid us proceeding for both calls of our `onComitted` event listener. The first time it's called, the timestamp will have also been updated and so the function will return. The second time its called, the function will proceed since the timestamp won't have changed again.

The thing is, I think the logic to check and clear the `showWelcomeBanner` and `showCounterMessaging` should also have the same "debouncing" effect. The first time the function is called, the settings will be true and are immediately cleared. The second time the function is called, the settings will have been cleared and the function will return. To check this, I've done a test where the `onComitted` listener logs the value of the settings, and I see they are `undefined` on the second call.

I also think it's suboptimal to continue listening for the `onComitted` event after the onboarding messaging is complete. So I have updated my changes here to:
1. Stop listening for `onComitted` after the onboarding is complete. I took the conservative (and simpler to reason about) approach of waiting until the next time the event listener fires.
2. Remove the `onBeforeNavigateTimeStamp` logic.
3. Restore the `settings.removeSetting(showWelcomeBanner)` logic to happen consistently.

This all seems to be working for me, testings are passing and I see the welcome banner. That said, any chance you could perform another test of the challenge flow like you did above, to confirm the banner is not displayed twice? That would save me setting up the whole SERP dev environment which I'm not familiar with.